### PR TITLE
[Mhlo] More ops and make Mhlo optional

### PR DESCRIPTION
**Pull request by Connor-XY**
This pull request:
1. Add option to the CMakeLists.txt so that Mhlo could be an optional library.
2. Add target as the parameter for decompose pass.
3. Add Mhlo lowering for Pow, Sqrt, Mul, MatMul, Conv, Argmax, Flatten, Shape, Transpose, Identity, Slice, Split, Squeeze, Unsqueeze, and Gather.

### CODE REVIEW

File: src/Conversion/CMakeLists.txt
Comment (sstamenova) [inline src/Conversion/CMakeLists.txt:5]: Nit: use two spaces for cmake files indentation

File: CMakeLists.txt
Comment (sstamenova) [inline CMakeLists.txt:12]: This list is alphabetic

**Comment by sstamenova** [comment]
Making MHLO optional is very exciting. Thank you!

**Comment by tungld** [comment]
@jenkins-droid test this please

File: src/Conversion/ONNXToMhlo/Math/MatMul.cpp
Comment (sorenlassen) [inline src/Conversion/ONNXToMhlo/Math/MatMul.cpp:57]: suggestion: define a helper function to map an IndexExpr to its literal value or -1

File: src/Conversion/ONNXToMhlo/Tensor/ArgMax.cpp
Comment (sorenlassen) [inline src/Conversion/ONNXToMhlo/Tensor/ArgMax.cpp:72]: message is not very informative, I think "Could not compute output shape" would be better

File: src/Conversion/ONNXToMhlo/NN/Conv.cpp
Comment (sorenlassen) [inline src/Conversion/ONNXToMhlo/NN/Conv.cpp:91]: do we know that all these things we call getLiteral() on are literals?

File: src/Conversion/ONNXToMhlo/NN/Conv.cpp
Comment (sorenlassen) [inline src/Conversion/ONNXToMhlo/NN/Conv.cpp:92]: do we know that inputShape[i + sparialOffset] is not -1?

**Review by sorenlassen — COMMENTED** [review]
I’m interested in learning about lowering so I read through the PR and made some comments, but I’m no expert and feel free to ignore if the comments are not helpful.

**Comment by sorenlassen** [comment]
@Connor-XY ^

File: test/mlir/conversion/onnx_to_mhlo/Math/Reduction.mlir
Comment (Connor-XY) [reply test/mlir/conversion/onnx_to_mhlo/Math/Reduction.mlir:1]: We decompose the softmax operator into primitive operators when the target is mhlo. We add the other two options to do shape inference and canonicalize the result

File: src/Conversion/ONNXToMhlo/Tensor/ArgMax.cpp
Comment (Connor-XY) [reply src/Conversion/ONNXToMhlo/Tensor/ArgMax.cpp:129]: The shape is dynamic, so we still need to use shape dialect to get the dim size of the output.

File: src/Conversion/ONNXToMhlo/Tensor/Flatten.cpp
Comment (Connor-XY) [reply src/Conversion/ONNXToMhlo/Tensor/Flatten.cpp:41]: Agree, I'll fix that.

File: src/Conversion/ONNXToMhlo/Tensor/ArgMax.cpp
Comment (Connor-XY) [reply src/Conversion/ONNXToMhlo/Tensor/ArgMax.cpp:98]: Agree, I'll fix that.

File: src/Conversion/ONNXToMhlo/Math/MatMul.cpp
Comment (Connor-XY) [reply src/Conversion/ONNXToMhlo/Math/MatMul.cpp:57]: I'll add the helper function

File: src/Conversion/ONNXToMhlo/Tensor/ArgMax.cpp
Comment (Connor-XY) [reply src/Conversion/ONNXToMhlo/Tensor/ArgMax.cpp:72]: will fix that

File: src/Conversion/ONNXToMhlo/NN/Conv.cpp
Comment (Connor-XY) [reply src/Conversion/ONNXToMhlo/NN/Conv.cpp:91]: The current conv only supports where padding and kernel size are known. We will add support for dynamic shape.

Comment (sorenlassen) [reply src/Conversion/ONNXToMhlo/NN/Conv.cpp:91]: I see

I suggest you add assertions to make it clear that pads, kernelShape, outputDims, and inputShape must all be known (unless this is already checked elsewhere)

File: src/Conversion/ONNXToMhlo/Tensor/Flatten.cpp
Comment (AlexandreEichenberger) [reply src/Conversion/ONNXToMhlo/Tensor/Flatten.cpp:40]: I would  first add the rank when negative, and then the axis should be from 0 inclusive to rank exclusive.

File: test/mlir/conversion/onnx_to_mhlo/Math/Reduction.mlir
Comment (AlexandreEichenberger) [reply test/mlir/conversion/onnx_to_mhlo/Math/Reduction.mlir:1]: I know we may not have set the best example, the preferred way is to test only one pass at a time, so that potential errors are easier to detect. I would then do 2 test files, one to test the first pass and one to do the second pass. 

We have often added `canonicalize` as a secondary pass as often the output without canonicalize is very redundant, so using this as a second pass is fine if it makes the test easier.

File: src/Conversion/ONNXToMhlo/Math/MatMul.cpp
Comment (AlexandreEichenberger) [reply src/Conversion/ONNXToMhlo/Math/MatMul.cpp:57]: This function already exists, check in IndexExpr.hpp
```C++
  // Helpers for list of IndexExpressions
  static void getShape(llvm::SmallVectorImpl<IndexExpr> &indexExprList,
      llvm::SmallVectorImpl<int64_t> &intDimList);
```

File: src/Conversion/ONNXToMhlo/NN/Conv.cpp
Comment (AlexandreEichenberger) [reply src/Conversion/ONNXToMhlo/NN/Conv.cpp:91]: My recollection is that we support dynamic images sizes, but expect the filter size to be known at compile time...
I will try to add some dyn size testing if its not already there.

**Comment by AlexandreEichenberger** [comment]
@jenkins-droid test this please

File: src/Conversion/ONNXToMhlo/NN/Conv.cpp
Comment (AlexandreEichenberger) [reply src/Conversion/ONNXToMhlo/NN/Conv.cpp:91]: Confirmed, in test/modellib/ConvModel and test/numerical/TestConv we are building/testing 2D convolutions that are dynamic in each of the NCHW dimensions.

**Comment by AlexandreEichenberger** [comment]
@jenkins-droid test this please

File: src/Conversion/ONNXToMhlo/NN/Conv.cpp
Comment (AlexandreEichenberger) [inline src/Conversion/ONNXToMhlo/NN/Conv.cpp:102]: Our shape inference compute left/right top/bottom padding. Is the mhlo padding different than onnx and thus justify recomputing some of it here?

**Review by AlexandreEichenberger — COMMENTED** [review]
I left some questions and comments in the text. 
I can see that there is a pattern that is starting to add a lot of code: if dimensions are constant, then use them, otherwise introduce a lot of shape ops to compute it.

We used a similar pattern at the beginning (before shape dialect was established I believe). It introduced a lot of redundant code, that was difficult to debug/upgrade as ONNX ops sometime change what is attributes vs what can become a dynamic input (with potentially dynamic dimensions).

This is why we introduced IndexExpr which can (depending on the context) use the same computation to either compute a shape number (nonnegative if known at compile time, -1 if dynamic) OR introduce computations to determine the shape.

I would generally encourage you to do so, possibly updating our IndexExpr to use shape operations if you don't tolerate our computations. I believe we use mostly arithmetic plus DimOp, which fetch the dimension out of a tensor. We have a slight complication in that we have accelerators that use maps for data layout, and we need to keep both the iteration space (original sizes of the tensors) as well as the mapped data size. We had to introduce some ops to handle this, maybe MLIR support has improved since then.

Basically, I am encouraging you to reuse our infra (and possibly upgrade it a bit) so that you don't have to generate a lot of complicated code to recompute the shape when dynamic using redundant code.

File: src/Conversion/ONNXToMhlo/NN/Conv.cpp
Comment (yaochengji) [reply src/Conversion/ONNXToMhlo/NN/Conv.cpp:102]: Let me help answer this question. `mhlo.convolution` is a generic N-D conv and its padding is of shape (N, 2), where padding[i] specifies lower and upper padding values of spatial dimension i. You can refer to [this](https://www.tensorflow.org/xla/operation_semantics#convwithgeneralpadding_convolution) for more details.

**Comment by yaochengji** [comment]
@AlexandreEichenberger do you think we need to add some milestone model tests here? Ex. resnet 18, bert-tiny.

**Comment by AlexandreEichenberger** [comment]
> @AlexandreEichenberger do you think we need to add some milestone model tests here? Ex. resnet 18, bert-tiny.

We currently test for a few benchmarks as part of our backend tests (run on all of our mandatory CIs): from inference_backend.py:

```
        "test_densenet121_cpu": {STATIC_SHAPE:{}},
        "test_inception_v1_cpu": {STATIC_SHAPE:{}},
        "test_resnet50_cpu": {STATIC_SHAPE:{}, DYNAMIC_SHAPE:{0:{-1}}},
        "test_shufflenet_cpu": {STATIC_SHAPE:{}},
        "test_squeezenet_cpu": {STATIC_SHAPE:{}},
        "test_vgg19_cpu": {STATIC_SHAPE:{}},
```

I am not aware on how to add tests there, we need to be a bit careful about not taking too long as some of our CIs may timeout.
If you believe that an additional benchmark should be here, I would suggest a distinct PR to add it.

File: src/Conversion/ONNXToMhlo/Math/Elementwise.cpp
Comment (Connor-XY) [reply src/Conversion/ONNXToMhlo/Math/Elementwise.cpp:142]: This is reasonable and we will add the comment

File: src/Conversion/ONNXToMhlo/Math/Elementwise.cpp
Comment (Connor-XY) [reply src/Conversion/ONNXToMhlo/Math/Elementwise.cpp:205]: This is answered above

File: src/Conversion/ONNXToMhlo/Math/MatMul.cpp
Comment (Connor-XY) [reply src/Conversion/ONNXToMhlo/Math/MatMul.cpp:62]: Yes, MHLO can handle -1 shapes. It seems that we need to figure out how to reuse IndexExpr and modify your previous implementation for MHLO conversion. We can use the getShape() function, and also add the new call in IndexExpr.

File: src/Conversion/ONNXToMhlo/NN/Conv.cpp
Comment (Connor-XY) [reply src/Conversion/ONNXToMhlo/NN/Conv.cpp:36]: I remember that when I tried to use shapehelper for sliceOp, fGetDenseVal failed since we only pass the operator pointer to the shapeHelper, but I couldn't remember clearly what the issue is.

File: src/Conversion/ONNXToMhlo/NN/Conv.cpp
Comment (Connor-XY) [reply src/Conversion/ONNXToMhlo/NN/Conv.cpp:84]: Yes, we can use the function to test if all index expressions are literals, and we need to modify IndexExpr so we can reuse the result from the shape helper.

**Comment by sstamenova** [comment]
@Connor-XY How difficult would it be to split out the change that makes mhlo optional? It seems like there's no concern about that particular aspect and it would be great to have sooner rather than later.

**Comment by yaochengji** [comment]
> @Connor-XY How difficult would it be to split out the change that makes mhlo optional? It seems like there's no concern about that particular aspect and it would be great to have sooner rather than later.

Hi @sstamenova ,

@Connor-XY gets a cold these days. I think making mhlo optional is not difficult.

File: src/Conversion/ONNXToMhlo/NN/Conv.cpp
Comment (Connor-XY) [reply src/Conversion/ONNXToMhlo/NN/Conv.cpp:102]: Due to ceil mode in ONNX, the padding can be the same even when the output shape differs when auto_pad is NOTSET. Therefore, we need to recalculate the padding. You could refer to @test_default_maxpoolsingleout_strides_nonunifpad and @test_default_maxpoolsingleout_strides_nonunifpad_ceil in both test/mlir/conversion/onnx_to_mhlo/NN/Pooling.mlir and test/mlir/onnx/onnx_shape_inference_maxpool.mlir, and see the difference.

File: src/Conversion/ONNXToMhlo/NN/Conv.cpp
Comment (Connor-XY) [reply src/Conversion/ONNXToMhlo/NN/Conv.cpp:84]: You could refer to the case below where we need to calculate the padding again. For simplicity we previously just calculate all paddings instead of reuse those in the shape helper.

**Comment by Connor-XY** [comment]
@AlexandreEichenberger I believe it is beneficial if we can reuse IndexExpr, but I currently have some concerns about this. First of all, it seems that in shape inference pass, when dimension is dynamic, all of the IndexExprs are QuestionmarkIndexExpr. Ideally we hope that the final result could be a Value and we generate the code after calling the shape helper so we can reuse the result in the lowering. This is different from what it is currently and I believe this requires substantial changes to the current design. What's more, I currently couldn't figure out a good way to differentiate Mhlo tensor and memref. I think it should be done wrt the scope or the rewriter, but couldn't think of a proper way to do that. We would like to ask for some suggestions about how we should modify IndexExpr so we could reuse your calculation while not affect the proper usage of the structure.

**Comment by AlexandreEichenberger** [comment]
> all of the IndexExprs are QuestionmarkIndexExpr.

and that is fine. When half the dims are constants, and the other are dynamic, the results will be a mixture of literals and question marks.

> final result could be a Value 

when doing shape inference, the only interesting outcome (short of symbolic analysis) is either constant lit or unknown. There is no need for values at this time. When lowering code, then we want the literal and/or the computations, all of which can be values. There are methods that transform index expressions in literal (individual or list of).

> to differentiate Mhlo tensor and memref. 

are MHLO tensors different from the MLIR tensors? I would look at how we use index expression now. During shape inference, for sure we use the tensors. During lowering, we have access to both unconverted inputs (tensors) or converted inputs (members). Not 100% sure which one we use.

I believe we want to avoid redundancy of code during shape inference and lowering. How we get there is less important, though we have invested in IndexExpr and it would be nice to adapt it to everyone's need. We can look together for solutions that works for your needs

**Comment by yaochengji** [comment]
> when doing shape inference, the only interesting outcome (short of symbolic analysis) is either constant lit or unknown.

@Connor-XY , agreed with @AlexandreEichenberger, it is shape inference, not shape reification, we only need shape calculation for the needs of operations like mhlo.dynamic_reshape.


> are MHLO tensors different from the MLIR tensors? 

@Connor-XY I don't think we need to take memref into consideration. All the inputs/outputs of MHLO and ONNX dialect are of type tensor.

**Comment by yaochengji** [comment]
Hi @AlexandreEichenberger , currently @Connor-XY 's internship in Bytedance is over so I'll take over this PR.

I have a question when extracting `Value` from `MemRefBoundsIndexCapture` or `IndexExpr`.
Ex. when converting onnx flattern op to mhlo, currently the code snippet handling the shape computation is as below:
```
Value flattenDimFirst = rewriter.create<arith::ConstantIndexOp>(loc, 1);
Value inputShape = rewriter.create<shape::ShapeOfOp>(loc, input);
for (int64_t i = 0; i < axis; i++) {
  Value dim = rewriter.create<shape::GetExtentOp>(loc, inputShape, i);
  flattenDimFirst =
      rewriter.create<shape::MulOp>(loc, flattenDimFirst, dim);
}
```

After changing to `IndexExpr` utility, I guess it should be as below:
```
MemRefBoundsIndexCapture XBounds(op.X());
IndexExpr mulDim = XBounds.get(0);
for (int64_t i = 1; i < axis; i++) {
  mulDim = mulDim * XBounds.get(1);
}
```

Besides, it seems I should add a correct `IndexExprScope` and modify the `MemRefBoundsIndexCapture::get` implementation code to make it works correct. Therefore, I think in the `Value` extraction scenario `IndexExpr` doesn't simplify the code very much. Could we keep code like this as it is?

**Comment by yaochengji** [comment]
Hi @AlexandreEichenberger, I rebased this branch and fixed some error after it. Could you take a look again?

**Comment by philass** [comment]
@jenkins-droid test this please

File: src/Conversion/ONNXToMhlo/Tensor/Transpose.cpp
Comment (philass) [inline src/Conversion/ONNXToMhlo/Tensor/Transpose.cpp:50]: I think this should be `has_value()`, and thats what causes the windows pipeline to fail.

`hasValue()` was deprecated in MLIR

Comment (yaochengji) [reply src/Conversion/ONNXToMhlo/Tensor/Transpose.cpp]: Fixed, thanks @philass

**Comment by philass** [comment]
@jenkins-droid test this please

**Review by AlexandreEichenberger — COMMENTED** [review]
Some work for the copyright comments.
This code still has a lot of "if constant do this, if dynamic do that". That is the pattern that our ShapeInference data structures avoid, as we use IndexExpression to compute either the dynamic value, or simplify to integer when all the values involved in the computations are constant.

My concern is that the code with the "if constant/dynamic" pattern is more complex to maintain long term. So I continue to advocate to migrate to that new code pattern.

Maybe you want to put this code down first, and then investigate if you can migrate to using our shape inference helper classes?

File: src/Conversion/ONNXToMhlo/Math/MatMul.cpp
Comment (yaochengji) [reply src/Conversion/ONNXToMhlo/Math/MatMul.cpp]: I only put the date this time. Because I don't know what process to go through to put the company name (ByteDance) here. It may take some time to figure out.

**Comment by yaochengji** [comment]
> My concern is that the code with the "if constant/dynamic" pattern is more complex to maintain long term. So I continue to advocate to migrate to that new code pattern.

If your concern is "if constant/dynamic" pattern makes things complex, we can treat all situation as dynamic in onnx->mhlo conversion and then a canonicalization pass will be added after that to simplify the static situations. BTW, this is how it implemented in TF. For IndexExpr, my concern is that it is not flexible enough to support arbitrary cases.


> Maybe you want to put this code down first, and then investigate if you can migrate to using our shape inference helper classes?

Yes, could we refine the code later.

File: src/Conversion/ONNXToMhlo/Math/MatMul.cpp
Comment (AlexandreEichenberger) [reply src/Conversion/ONNXToMhlo/Math/MatMul.cpp]: Then just leave your company out... but it does not seems to be right to add `IBM Authors` when you are not employed by IBM, if you see what I mean.

**Comment by AlexandreEichenberger** [comment]
> Besides, it seems I should add a correct IndexExprScope and modify the MemRefBoundsIndexCapture::get implementation code to make it works correct. Therefore, I think in the Value extraction scenario IndexExpr doesn't simplify the code very much. Could we keep code like this as it is?

The issue is not so much using index expression vs adding code directly, it is more about reusing the shape inference helpers (or of they are not used, maybe migrating the current code into a shape helper so that it may be reused).

Unless the computations are distinct because you are targeting MHLO, our goal should be to reuse shape inference as much as possible and avoid "if constant/dynamic" branches.

File: src/Conversion/ONNXToMhlo/Math/MatMul.cpp
Comment (yaochengji) [reply src/Conversion/ONNXToMhlo/Math/MatMul.cpp]: I didn't add `IBM Authros`, either. Only `Copyright 2022` currently.

**Comment by yaochengji** [comment]
fix some conflicts.

@AlexandreEichenberger do you have any other comments? If not, could this PR be merged?

**Comment by yaochengji** [comment]
> fix some conflicts.
> 
> @AlexandreEichenberger do you have any other comments? If not, could this PR be merged?

@tungld could you help take a look at this? (I guess @AlexandreEichenberger is on leave these days)

**Comment by yaochengji** [comment]
@tungld the test passed. Is there any work we should do on this PR?

**Comment by tungld** [comment]
@yaochengji the patch looks good to me. However, still would like to hear the final decision from @AlexandreEichenberger as he has followed the patch since its inception.

**Review by AlexandreEichenberger — APPROVED** [review]
LGTM, urging future PR to rely more on existing shape inference framework to reduce redundancy.

**Comment by yaochengji** [comment]
> LGTM, urging future PR to rely more on existing shape inference framework to reduce redundancy.

Sure, this will be our next PR.

**Comment by yaochengji** [comment]
@AlexandreEichenberger As it is already approved, could you merge this PR after CI passes?

**Comment by AlexandreEichenberger** [comment]
> As it is already approved, could you merge this PR after CI passes?

will do

**Pull request merged**
